cli/azd/internal/vsrpc/environment_service_deploy.go

```diff
@@ -0,0 +1,95 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package vsrpc
+
+import (
+	"context"
+
+	"github.com/azure/azure-dev/cli/azd/cmd/actions"
+	"github.com/azure/azure-dev/cli/azd/internal"
+	"github.com/azure/azure-dev/cli/azd/internal/cmd"
+	"github.com/azure/azure-dev/cli/azd/pkg/ioc"
+)
+
+// DeployAsync is the server implementation of:
+// ValueTask<Environment> DeployAsync(Session, string, IObserver<ProgressMessage>, CancellationToken)
+//
+// While it is named simply `DeployAsync`, it behaves as if the user had run `azd provision` and `azd deploy`.
+func (s *environmentService) DeployAsync(
+	ctx context.Context, sessionId Session, name string, observer IObserver[ProgressMessage],
+) (*Environment, error) {
+	session, err := s.server.validateSession(ctx, sessionId)
+	if err != nil {
+		return nil, err
+	}
+
+	session.sessionMu.Lock()
+	defer session.sessionMu.Unlock()
+
+	outputWriter := &lineWriter{
+		next: &messageWriter{
+			ctx:      ctx,
+			observer: observer,
+		},
+	}
+
+	session.outWriter.AddWriter(outputWriter)
+	defer session.outWriter.RemoveWriter(outputWriter)
+
+	provisionFlags := cmd.NewProvisionFlagsFromEnvAndOptions(
+		&internal.EnvFlag{
+			EnvironmentName: name,
+		},
+		&internal.GlobalCommandOptions{
+			Cwd:      session.rootPath,
+			NoPrompt: true,
+		},
+	)
+
+	deployFlags := cmd.NewDeployFlagsFromEnvAndOptions(
+		&internal.EnvFlag{
+			EnvironmentName: name,
+		},
+		&internal.GlobalCommandOptions{
+			Cwd:      session.rootPath,
+			NoPrompt: true,
+		},
+	)
+
+	session.container.MustRegisterScoped(func() internal.EnvFlag {
+		return internal.EnvFlag{
+			EnvironmentName: name,
+		}
+	})
+
+	ioc.RegisterInstance[*cmd.ProvisionFlags](session.container, provisionFlags)
+	ioc.RegisterInstance[*cmd.DeployFlags](session.container, deployFlags)
+	ioc.RegisterInstance[[]string](session.container, []string{})
+
+	session.container.MustRegisterNamedTransient("provisionAction", cmd.NewProvisionAction)
+	session.container.MustRegisterNamedTransient("deployAction", cmd.NewDeployAction)
+
+	var c struct {
+		deployAction    actions.Action `container:"name"`
+		provisionAction actions.Action `container:"name"`
+	}
+
+	if err := session.container.Fill(&c); err != nil {
+		return nil, err
+	}
+
+	if _, err := c.provisionAction.Run(ctx); err != nil {
```

Comment:
Just trying to think about the system behavior... when a provision + deploy runs, we hold on to the `sessionMu` lock for the full duration. Other requests with the same session would end up blocking. Does this mean in-practice, VS ends up sending other session IDs to avoid the scenario?

Reply:
> . Other requests with the same session would end up blocking.

Yup - it's not great :-/, but this was the quickest way to get things off the ground with a level of confidence I'm comfortable with.

> Does this mean in-practice, VS ends up sending other session IDs to avoid the scenario?

Nope - it means in practice VS has a bad experience (it takes a long time for data to come back and progress wheels are just spinning) - we need to figure out how to do better.  The first baby step is to observe cancellation in a reasonable number of locations (strategic places in RefreshAsync and DeployAsync would be ideal) so we can at least abort the operation at safe points somewhat eagerly.

The next step after that is more finer grained locking - to be honest, I feel like if we refactor things such that the server builds its root context instead of it being passed in and we register everything as transients (with perhaps a few small things as singletons to get good caching on app host manifest fetches) then we might only need to lock around the calls to `container.Fill` and things will just all work nicely.

#3287 tracks the complete fix - I hope to tackle it soon.

Reply:
I also added this comment to `serverSession`:

```go
// TODO(azure/azure-dev#3287): Today we lock the session at the start of each RPC and hold it for the entire lifetime of the
// RPC. In practice this can be problematic because DeployAsync and RefreshEnvironmentAsync can take a long time to run.
// Ideally we would do finer grained locking the session itself and know that the rest of the system is safe to run
// concurrently. To do this we need to get better disciplined about what state we store and how we access it, so for now
// we just do very coarse grained locking.
```